### PR TITLE
remove logging hack

### PR DESCRIPTION
**Pull request by stanlyDoge**
Removing hack which used EndVars to restart the pods when the logging configuration has been changed. Now it uses `internalPatch` to determine if the resource (in this case CM) has been changed or not.

### CODE REVIEW

**Comment by scholzj** [comment]
No, its not OK to restart all Kafkas and Zoos just because CO restarted. They need to be rock stable :-o.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java [hunk 1]
```diff
@@ -329,6 +329,7 @@ public String parseLogging(Logging logging, ConfigMap externalCm) {
 
         } else {
             // field is not in the cluster CM
+            this.setLogging(new InlineLogging());
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java]: this isn't the right place to do this, I think. If the default is an `InlineLogging` that should be set in `fromCrd()`. After `fromCrd()` is call we should treat the `KafkaCluster` (or whatever subclass) as immutable.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 2]
```diff
@@ -64,6 +66,8 @@
     private final PvcOperator pvcOperations;
     private final DeploymentOperator deploymentOperations;
     private final ConfigMapOperator configMapOperations;
+    protected ConfigMap oldKafkaCM = null;
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: The operator cannot be stateful like this.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 3]
```diff
@@ -269,11 +273,20 @@ Secret brokersInternalSecret() {
...
+                        log.debug("Kafka configmap change detected. Restarting pods");
+                        ss.getSpec().getTemplate().getMetadata().getAnnotations().put(ANNOTATION_GENERATION, "-1");
+                    } else {
+                        log.debug("Kafka fresh start");
+                    }
+
                     KafkaClusterDescription desc =
                             new KafkaClusterDescription(kafka, kafka.generateService(), kafka.generateHeadlessService(),
-                                    logAndMetricsConfigMap, kafka.generateStatefulSet(isOpenShift),
+                                    logAndMetricsConfigMap, ss,
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: I think what you need to do is use the return value from the reconciliation of the log and metrics config map to inform whether a roll is necessary by the `maybeRollPods()`. I think it would be OK for `maybeRollPods()` to take a `boolean forceRoll` parameter to do that.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ConfigMapOperator.java [hunk 4]
```diff
@@ -29,4 +32,22 @@ public ConfigMapOperator(Vertx vertx, KubernetesClient client) {
     protected MixedOperation<ConfigMap, ConfigMapList, DoneableConfigMap, Resource<ConfigMap, DoneableConfigMap>> operation() {
         return client.configMaps();
     }
+
+    @Override
+    protected Future<ReconcileResult<ConfigMap>> internalPatch(String namespace, String name, ConfigMap current, ConfigMap desired) {
+        try {
+            TreeMap a = new TreeMap(current.getData());
+            TreeMap b = new TreeMap(desired.getData());
+            if (a.toString().equals(b.toString())) {
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ConfigMapOperator.java]: You can just use [`.equals()` on the maps](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#equals-java.lang.Object-), no need to convert them to Strings. This also means there's no need to copy the maps into `TreeMaps` for consistent ordering.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ConfigMapOperator.java [hunk 5]
```diff
@@ -29,4 +32,22 @@ public ConfigMapOperator(Vertx vertx, KubernetesClient client) {
...
+                log.debug("{} {} in namespace {} has not been patched because resources are equal", resourceKind, name, namespace);
+                return Future.succeededFuture(ReconcileResult.noop());
+            }
+            log.debug("{} {} in namespace {} has been patched", resourceKind, name, namespace);
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ConfigMapOperator.java]: You can just call `super.internalPatch`. No need to reimplement the same code here.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java [hunk 6]
```diff
@@ -105,8 +106,8 @@ public StatefulSetOperator(Vertx vertx, KubernetesClient client, long operationT
             }).compose(ignore -> {
                 Future del = podOperations.waitFor(namespace, name, pollingIntervalMs, timeoutMs, (ignore1, ignore2) -> {
                     // predicate - changed generation means pod has been updated
-                    int newGen = getPodGeneration(podOperations.get(namespace, podName));
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java]: Is `getPodGeneration()` still used somewhere else? If not then you should remove it.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java [hunk 7]
```diff
@@ -290,4 +291,31 @@ protected StatefulSetDiff revertStorageChanges(StatefulSet current, StatefulSet
...
+    }
+
+    private static String getPodUid(Pod resource) {
+        if (resource == null) {
+            return NO_UID;
+        }
+        return getUid(resource.getMetadata());
+    }
+
+    private static String getUid(ObjectMeta objectMeta) {
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java]: I don't think there's much benefit in having both getPodUid() and getUid(). They could be merged into a single `getUid(HasMetadata)` method.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/MockKube.java [hunk 8]
```diff
@@ -51,6 +51,7 @@
 import io.fabric8.kubernetes.client.dsl.Resource;
 import io.fabric8.kubernetes.client.dsl.RollableScalableResource;
 import io.fabric8.kubernetes.client.dsl.ScalableResource;
+import org.apache.commons.lang.RandomStringUtils;
```
Comment (tombentley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/MockKube.java]: No need to add new dependency: You can just use `java.util.UUID`, I suspect

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/AbstractResourceOperatorTest.java [hunk 9]
```diff
@@ -106,7 +106,7 @@ public void createWhenExistsIsAPatch(TestContext context, boolean cascade) {
             }
             assertTrue(ar.succeeded());
             verify(mockResource).get();
-            verify(mockResource).patch(any());
+            //verify(mockResource).patch(any());
```
Comment (tombentley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/AbstractResourceOperatorTest.java]: Why is this necessary?

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperatorTest.java [hunk 10]
```diff
@@ -138,9 +139,14 @@ protected boolean isPodUpToDate(StatefulSet ss, String podName) {
             protected Future<Integer> getGeneration(String namespace, String podName) {
                 return Future.succeededFuture(1);
             }
+
+            @Override
+            protected Future<String> getUid(String namespace, String podName) {
+                return Future.succeededFuture(RandomStringUtils.randomAlphanumeric(12));
```
Comment (tombentley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperatorTest.java]: `UUID` again

File: examples/kafka/kafka-ephemeral.yaml [hunk 11]
```diff
@@ -38,4 +38,7 @@ spec:
       type: ephemeral
     metrics:
       lowercaseOutputName: true
+    logging:
+      type: external
+      name: karel
```
Comment (tombentley) [inline examples/kafka/kafka-ephemeral.yaml]: `karel`?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java (hunk 6 above)
Comment (stanlyDoge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java]: It's still used in `isPodUpToDate()`. Do you think i should refactor `isPodUpToDate` to use UID as well?

Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java]: I was just checking it *was* still used. Don't change `isPodUpToDate()`.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/AbstractResourceOperatorTest.java (hunk 9 above)
Comment (stanlyDoge) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/AbstractResourceOperatorTest.java]: `ConfigMapOperatorTest` runs this `createWhenExistsIsAPatch()` but current and desired CMs are equal. It expects the `patch()` to be called but nothing has changed and `patch()` is not called.

Comment (tombentley) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/AbstractResourceOperatorTest.java]: Mmm, Perhaps it's always do the `patch()` in the `ConfigMapOperator.internalPatch()`, and evaluate whether the CM `data` changed after the call to the `super.internalPatch()`. The reason is that by not calling `patch()` we're not going to update the CM for other things than `data` like `metadata.labels`. Doing that would also mean you could reinstate this `verify()` call. It *would* mean that what we were returning from `ConfigMapOperator.internalPatch()` could sometimes be a no op when the patch wasn't actually a no op, but that's not a distinction that matters currently, so I think it would be acceptable.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ConfigMapOperator.java [hunk 12]
```diff
@@ -29,4 +30,18 @@ public ConfigMapOperator(Vertx vertx, KubernetesClient client) {
     protected MixedOperation<ConfigMap, ConfigMapList, DoneableConfigMap, Resource<ConfigMap, DoneableConfigMap>> operation() {
         return client.configMaps();
     }
+
+    @Override
+    protected Future<ReconcileResult<ConfigMap>> internalPatch(String namespace, String name, ConfigMap current, ConfigMap desired) {
+        try {
+            if (current.getData().equals(desired.getData())) {
+                log.debug("{} {} in namespace {} has not been patched because resources are equal", resourceKind, name, namespace);
+                return Future.succeededFuture(ReconcileResult.noop());
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ConfigMapOperator.java]: This still won't patch when the `data` are equal. That's wrong. Consider what happens if the user adds a label to their `Kafka` resource. What should happen is all the dependent resources get the same label. If we don't call `internalPatch()` the CM won't have its label changed. So you should always call `super.internalPatch()`, but use the `data` maps to determine whether to _say_ the result was a no op. It's a lie, but I think it's acceptable (at least at the moment).

**Comment by scholzj** [comment]
I do not see any changes for Kafka Connect and TO ... how does the update work for them?

**Comment by stanlyDoge** [comment]
@scholzj I've added functionality for Kafka Connect, Kafka Connect S2I and TO. Can you review these changes?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectS2ICluster.java [hunk 13]
```diff
@@ -28,13 +28,16 @@
...
+import java.util.Map;
+
 public class KafkaConnectS2ICluster extends KafkaConnectCluster {
 
     // Kafka Connect S2I configuration
     protected String sourceImageBaseName = DEFAULT_IMAGE.substring(0, DEFAULT_IMAGE.lastIndexOf(":"));
     protected String sourceImageTag = DEFAULT_IMAGE.substring(DEFAULT_IMAGE.lastIndexOf(":") + 1);
     protected String tag = "latest";
     protected boolean insecureSourceRepository = false;
+    static Map<String, String> configMapS2I;
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectS2ICluster.java]: Why is this `static`?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ConfigMapOperator.java (hunk 12 above)
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ConfigMapOperator.java]: This still needs fixing.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectS2ICluster.java (hunk 13 above)
Comment (stanlyDoge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectS2ICluster.java]: This map is used to save "previous" configuration which is later compared to the current configuration.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectS2ICluster.java [hunk 14]
```diff
@@ -128,6 +131,11 @@ public DeploymentConfig generateDeploymentConfig() {
         return dc;
     }
 
+    public DeploymentConfig generateDeploymentConfig(Map annotations) {
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectS2ICluster.java]: Why overload like this? Why not just change the existing `generateDeploymentConfig()` method, since I think it has only a single caller anyway, right?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectS2ICluster.java [hunk 15]
```diff
@@ -28,13 +28,16 @@
...
+import java.util.Map;
+
 public class KafkaConnectS2ICluster extends KafkaConnectCluster {
 
     // Kafka Connect S2I configuration
     protected String sourceImageBaseName = DEFAULT_IMAGE.substring(0, DEFAULT_IMAGE.lastIndexOf(":"));
     protected String sourceImageTag = DEFAULT_IMAGE.substring(DEFAULT_IMAGE.lastIndexOf(":") + 1);
     protected String tag = "latest";
     protected boolean insecureSourceRepository = false;
+    static String configMapS2I;
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectS2ICluster.java]: Again, this should not be necessary.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 16]
```diff
@@ -278,7 +308,10 @@ TopicOperator topicOperator() {
             return this.topicOperator;
         }
 
-        Deployment deployment() {
+        Deployment deployment(boolean forceRestart) {
+            if (forceRestart) {
+                this.deployment.getSpec().getTemplate().getMetadata().getAnnotations().put("strimzi.io/logging", this.metricsAndLogsConfigMap.getData().toString());
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: You're just using this to force a rolling update of the deployment, aren't you?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ConfigMapOperator.java [hunk 17]
```diff
@@ -29,4 +33,41 @@ public ConfigMapOperator(Vertx vertx, KubernetesClient client) {
...
+        } catch (Exception e) {
+            log.error("Caught exception while patching {} {} in namespace {}", resourceKind, name, namespace, e);
+            return Future.failedFuture(e);
+        }
+    }
+
+    private boolean compareObjects(Object a, Object b) {
+        if (a instanceof Map || b instanceof Map) {
+            if (a == null && b != null) {
+                if (((HashMap) b).size() == 0) {
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ConfigMapOperator.java]: You've checked that b is a Map, not a `HashMap`, so this could CCE.
